core/domain/platform_parameter_domain_test.py

```diff
@@ -81,3 +82,1099 @@ def test_to_dict(self):
+            'data_type': 'bool',
+            'rules': [
+                {
+                    'filters': [
+                        {'type': 'server_mode', 'conditions': [('=', 'dev')]}
+                    ],
+                    'value_when_matched': False
+                }
+            ],
+            'rule_schema_version': (
+                feconf.CURRENT_PLATFORM_PARAMETER_RULE_SCHEMA_VERSION),
+            'default_value': False,
+            'metadata': {
+                'is_feature': True,
+                'feature_stage': 'dev',
+            }
+        })
+        parameter.validate()
+
+    def test_validate_feature_with_invalid_type_raises_exception(self):
+        parameter = parameter_domain.PlatformParameter.from_dict({
+            'name': 'parameter_a',
+            'description': 'for test',
+            'data_type': 'string',
+            'rules': [],
+            'rule_schema_version': (
+                feconf.CURRENT_PLATFORM_PARAMETER_RULE_SCHEMA_VERSION),
+            'default_value': '111',
+            'metadata': {
+                'is_feature': True,
+                'feature_stage': 'dev',
+            }
+        })
+        with self.assertRaisesRegexp(
+            utils.ValidationError,
+            'Data type of feature flags must be bool, got \'string\' instead'):
+            parameter.validate()
+
+    def test_validate_feature_with_invalid_stage_raises_exception(self):
+        parameter = parameter_domain.PlatformParameter.from_dict({
+            'name': 'parameter_a',
+            'description': 'for test',
+            'data_type': 'bool',
+            'rules': [],
+            'rule_schema_version': (
+                feconf.CURRENT_PLATFORM_PARAMETER_RULE_SCHEMA_VERSION),
+            'default_value': False,
+            'metadata': {
+                'is_feature': True,
+                'feature_stage': 'Invalid',
+            }
+        })
+        with self.assertRaisesRegexp(
+            utils.ValidationError, 'Invalid feature stage, got \'Invalid\''):
+            parameter.validate()
+
+    def test_validate_feature_with_no_mode_filter_raises_exception(self):
+        parameter = parameter_domain.PlatformParameter.from_dict({
+            'name': 'parameter_a',
+            'description': 'for test',
+            'data_type': 'bool',
+            'rules': [
+                {
+                    'filters': [],
+                    'value_when_matched': True
+                }
+            ],
+            'rule_schema_version': (
+                feconf.CURRENT_PLATFORM_PARAMETER_RULE_SCHEMA_VERSION),
+            'default_value': False,
+            'metadata': {
+                'is_feature': True,
+                'feature_stage': 'dev',
+            }
+        })
+        with self.assertRaisesRegexp(
+            utils.ValidationError, 'must have a server_mode filter'):
+            parameter.validate()
+
+    def test_validate_dev_feature_for_inappropriate_env_raises_exception(self):
```

Comment:
Here & elsewhere: be specific on the conditions (what env the parameter is in vs. the context) to make it much clearer what is actually 'inappropriate'.

Reply:
Done.